### PR TITLE
Vat termination phase3: throw away on-disk remnants of dead vats

### PR DESCRIPTION
**Pull request by FUDCo**

### CODE REVIEW

File: packages/SwingSet/src/kernel/state/kernelKeeper.js [hunk 1]
```diff
@@ -432,21 +432,42 @@ export default function makeKernelKeeper(storage) {
...
-      const kpid = storage.get(k);
-      const p = getKernelPromise(kpid);
-      if (p.state === 'unresolved' && p.decider === vatID) {
-        result.push(kpid);
+      if (k.startsWith(koPrefix)) {
+        const koid = storage.get(k);
+        const ownerKey = `${koid}.owner`;
+        const ownerVat = storage.get(ownerKey);
+        if (ownerVat === vatID) {
+          storage.set(ownerKey, 'none');
```
Comment (warner) [inline packages/SwingSet/src/kernel/state/kernelKeeper.js]: Let's add a note that this should always be the case: the `void` is `o+NN` (not `o-NN`), which means the vat exported the object (not imported) and allocated the `void` number, which means the vat owns it.

(this is not the case for promises, which can be exported/allocated by a vat but be decided by somebody else)

File: packages/SwingSet/src/kernel/state/kernelKeeper.js [hunk 2]
```diff
@@ -432,21 +432,42 @@ export default function makeKernelKeeper(storage) {
...
     }
-    return result;
+    // TODO: deleting entries from the dynamic vat IDs list requires a linear
+    // scan; this collection ought to be represented in a way that makes it
+    // efficient to remove an entry from it, though this should be OK as long as
+    // we keep the list short.
+    const DYNAMIC_IDS_KEY = 'vat.dynamicIDs';
+    const oldDynamicVatIDs = JSON.parse(getRequired(DYNAMIC_IDS_KEY));
+    const newDynamicVatIDs = oldDynamicVatIDs.filter(v => v !== vatID);
+    storage.set(DYNAMIC_IDS_KEY, JSON.stringify(newDynamicVatIDs));
```
Comment (warner) [inline packages/SwingSet/src/kernel/state/kernelKeeper.js:490]: we should probably sort this list before writing it back, it needs to be deterministic

the same "lexicographic is easier to do in JS, but numeric would be nicer for a spec document and in future non-JS languages" issue is present for this sort. Use lexicographic for now, but we need to revisit before we lock down the consensus spec.

File: packages/SwingSet/src/kernel/kernel.js [hunk 3]
```diff
@@ -39,6 +39,9 @@ function makeError(s) {
   return harden({ body: JSON.stringify(s), slots: [] });
 }
 
+const VAT_TERMINATION_ERROR = makeError('vat terminated');
+const UNKNOWN_VAT_ERROR = makeError('unknown vat');
```
Comment (warner) [inline packages/SwingSet/src/kernel/kernel.js]: I don't think userspace should ever be able to see this "unknown vat" error, and the fact that it exists makes me worried. They should only ever be able to see "vat terminated", and it shouldn't matter whether their message was early (pipelined to some promise that the kernel took over as executor of the dead vat's estate) or late (sent to an object-id whose owner is now "none"), they should always see the same error message.

File: packages/SwingSet/src/kernel/kernel.js [hunk 4]
```diff
@@ -322,7 +325,7 @@ export default function buildKernel(kernelEndowments, kernelOptions = {}) {
     const { type } = parseKernelSlot(target);
     if (type === 'object') {
       const vatID = kernelKeeper.ownerOfKernelObject(target);
-      insistVatID(vatID);
+      vatID === 'none' || insistVatID(vatID);
```
Comment (warner) [inline packages/SwingSet/src/kernel/kernel.js]: `insistVatID` was changed to tolerate `none`, is this bypass also necessary?

(I'm on the fence as to whether we add the null type to the definition of a VatID, which has one set of problems, or effectively create a new "vatID or None" type like this suggests)

**Review by warner — CHANGES_REQUESTED** [review]
Looks good! Let's chat about the "unknown vat" error a bit before landing.

File: packages/SwingSet/src/kernel/state/kernelKeeper.js (hunk 2 above)
Comment (FUDCo) [reply packages/SwingSet/src/kernel/state/kernelKeeper.js:490]: In this case sorting should not be needed -- essentially all we're doing is removing an element from an array, which will preserve the existing order of the elements that were already there.

File: packages/SwingSet/src/kernel/kernel.js (hunk 4 above)
Comment (FUDCo) [reply packages/SwingSet/src/kernel/kernel.js]: Yes, I think that check was a byproduct of churn during development.

File: packages/SwingSet/src/kernel/state/kernelKeeper.js (hunk 2 above)
Comment (FUDCo) [reply packages/SwingSet/src/kernel/state/kernelKeeper.js:490]: Also, prior to deleting any entries from this list, the order is not lexicographic anyway.  The ordering is the temporal order that the vats were created.  Given that the computation which creates the vats is deterministic, the ordering of the list must be as well.

The lexicographic ordering thing had to do with multiple entries in the kernel DB being sequenced lexicographically by key, but this list is always stored as a single database entry with a single key, `vat.dynamicIDs`.

**Pull request merged**